src/hostname/hostnamed.c

```diff
@@ -218,8 +219,9 @@ static int get_hardware_serial(char **ret) {
         if (r < 0)
                 return log_debug_errno(r, "Failed to open /sys/class/dmi/id device, ignoring: %m");
 
-        (void) sd_device_get_sysattr_value(device, "product_serial", &s);
-        if (isempty(s))
+        if (dmi_data)
+                (void) sd_device_get_sysattr_value(device, dmi_data, &s);
+        if (!s && strcmp(dmi_data,"product_serial"))
```

Comment:
given this is now a generic helper, do the fallback in `get_hardware_serial()` below instead